### PR TITLE
ZMS-261 getImapUid support

### PR DESCRIPTION
**Pull request by fatkudu**

### CODE REVIEW

File: store/src/java/com/zimbra/qa/unittest/TestZClient.java [hunk 1]
```diff
@@ -29,10 +33,10 @@
 import java.util.List;
 import java.util.Set;
 
-import junit.framework.Assert;
-import junit.framework.TestCase;
-
 import org.apache.commons.io.IOUtils;
+import org.junit.After;
```
Comment (grishick) [inline store/src/java/com/zimbra/qa/unittest/TestZClient.java]: This seems like an unnecessary change that causes a lot more non-functional changes throughout the rest of the file. I don't think it achieves anything other then making the diff very big.

File: client/src/java/com/zimbra/client/ZMessage.java [hunk 2]
```diff
@@ -561,13 +566,33 @@ public MailItemType getMailItemType() {
...
     @Override
     public int getImapUid() {
-        throw new UnsupportedOperationException("ZMessage method not supported yet");
+        if ((null != imapUid) && (imapUid >= 0)) {
+            return imapUid;
+        }
+        ZimbraLog.mailbox.debug("ZMessage getImapUid() - regetting UID");
+        ZMessage zm = null;
+        try {
+            zm = getMailbox().getMessageById(mId);
```
Comment (fatkudu) [inline client/src/java/com/zimbra/client/ZMessage.java]: The cache lookup misses if different things were requested - I need another look to check that the lookup fails if Uid wasn't asked for before.

File: client/src/java/com/zimbra/client/ZContact.java [hunk 3]
```diff
@@ -437,8 +443,31 @@ public InputStream getContentStream() throws ServiceException {
...
+        if ((null != imapUid) && (imapUid >= 0)) {
+            return imapUid;
+        }
+        if (isGalContact()) {
+            return 0;  /* TODO: is this the best thing to do? */
+        }
+        ZimbraLog.mailbox.debug("ZContact getImapUid() - regetting UID");
+        ZContact zc = null;
+        try {
+            zc = getMailbox().getContact(mId);
```
Comment (fatkudu) [inline client/src/java/com/zimbra/client/ZContact.java]: I assumed this worked the same as for messages but I don't think it does.  Needs more research.

Comment (fatkudu) [inline client/src/java/com/zimbra/client/ZContact.java]: I think this code is correct.  Added tests that exercise it.

File: client/src/java/com/zimbra/client/ZMessage.java (hunk 2 above)
Comment (fatkudu) [inline client/src/java/com/zimbra/client/ZMessage.java]: Added tests that exercise the code.  I think there are instances when it may be needed.

File: store/src/java/com/zimbra/qa/unittest/TestZClient.java (hunk 1 above)
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestZClient.java]: Changes now much smaller.  Could do with JUnit 4 changes at some point - but probably better on the develop branch.

**Comment by fatkudu** [comment]
Getting a couple of failures in TestRemoteImapShared -  
`testListContactsContents` fails because:
Caused by: java.lang.UnsupportedOperationException: ZContact method not supported yet
    at com.zimbra.client.ZContact.getContentStream(ZContact.java:443)
`testUidCopy` fails because although it claims to succeed, some info is missing.  I think I saw that Greg had taken on an Imap copy fix, so I'm hoping it will be addressed by that.  Otherwise can look at it later.

**Comment by grishick** [comment]
I pushed a small commit to address ZContact::getContentStream(). You can try rebasing or just plug this simple implementation:
`    @Override
    public InputStream getContentStream() throws ServiceException {
        return mMailbox.getRESTResource(String.format("?id=%s", this.getId()));
    }`
